substrate/client/network/src/service.rs

```diff
@@ -467,7 +467,7 @@ where
 							crate::MAX_CONNECTIONS_ESTABLISHED_INCOMING,
 						)),
 				)
-				.substream_upgrade_protocol_override(upgrade::Version::V1Lazy)
+				.substream_upgrade_protocol_override(upgrade::Version::V1)
```

Comment:
This is the version of the multistream-select protocol being used. This is the protocol used for negotiating the protocol that's going to be used on a newly-opened substream. For V1, it sends a message to the peer with the proposed protocol name and awaits a response. If the protocol name was confirmed, we get on with our lives. If not, we try other protocols until we reach a common denominator. If we cannot, we just close the substream (but the connection stays alive, since other protocols shouldn't be affected).

V1Lazy is a wire-compatible optimisation of the multistream-select protocol that rust-libp2p implements (it's not in the actual protocol spec). This version is optimistic that the remote will accept our first protocol so it sends out the protocol negotiation message as well as the actual protocol payload in a single network message (by writing to the socket one time instead of first waiting for the protocol confirmation).
Now, there's a bug in the version of rust-libp2p that we're using currently, which occurs if we use V1Lazy. If the protocol cannot be negotiated, rust-libp2p kills the entire connection. For this PR, if the peer is not upgraded to use the new chunk fetching protocol, we won't get a chance to try the older request type, because the connection would have closed on us.

So that's why we need to use V1, so that this PR's req-response backwards compatibility will work.

Now, one may ask, why not upgrade rust-libp2p to a version that fixes this. The answer is, there's work underway to do that, but it's a tremendous effort: https://github.com/paritytech/polkadot-sdk/pull/1631.
Moreover, it's still not clear whether the latest libp2p version fixes this, because on the upgrade PR, V1 had to be used to fix a very similar bug.

I spoke with Aaro and he agreed that this is fine to do until we manage to upgrade, test (and possibly fix) the latest rust-libp2p version. litep2p will be designed to accomodate for this new use case

Reply:
CC: @altonen